main.cpp

```diff
@@ -92,6 +92,86 @@ good to go!
  */
 
 #include <iostream>
+
+struct FloatType
+{
+    float add( float lhs, float rhs )
+    {
+        return lhs + rhs;
+    }
+    float subtract( float lhs, float rhs )
+    {
+        return lhs - rhs;
+    }
+    float multiply( float lhs, float rhs )
+    {
+        return lhs * rhs;
+    }
+    float divide( float lhs, float rhs )
+    {
+        if(rhs == 0.0f)
+        {
+            std::cout << std::endl;
+            std::cout << "warning, floating point division by zero returns 'inf' !" << std::endl;
+        }
+        return lhs / rhs;
+    }
+};
+
+struct DoubleType
+{
+    double add( double lhs, double rhs )
+    {
+        return lhs + rhs;
+    }
+    double subtract( double lhs, double rhs )
+    {
+        return lhs - rhs;
+    }
+    double multiply( double lhs, double rhs )
+    {
+        return lhs * rhs;
+    }
+    double divide( double lhs, double rhs )
+    {
+        if(rhs == 0.0)
+        {
+            std::cout << std::endl;
+            std::cout << "warning, floating point division by zero returns 'inf' !" << std::endl;
+        }
+        return lhs / rhs;
+    }
+};
+
+struct IntType
+{
+    int add( int lhs, int rhs )
+    {
+        return lhs + rhs;
+    }
+    int subtract( int lhs, int rhs )
+    {
+        return lhs - rhs;
+    }
+    int multiply( int lhs, int rhs )
+    {
+        return lhs * rhs;
+    }
+    int divide( int lhs, int rhs )
+    {
+        if(rhs == 0)
+        {
+            std::cout << "error, integer division by zero will crash the program!" << std::endl;
+            std::cout << "returning lhs" << std::endl;
+            return lhs;
+        }
+        else 
+        {
+            return lhs / rhs;
+        }
+    }
```

Comment:
If you find yourself writing code like this:
```
int Type::someFunction()
{
    if( someExpression )
    {
        someStatement();
        return aValue;
    } 
    else 
    {
        return anotherValue;
    }
}
```
You do not need the `else` block.  Here's why:
`return` statements are the last thing that **ever happens** in a function.
When the `if()` expression is evaluated as `true` and the block is entered, the `return aValue;` is the last line of code to be executed before the function exits.
but when the `if()` expression evaluates as `false` and the block is **NOT** entered, the next thing that happens is `return anotherValue;`
you have no code after the `else` block that needs executing, since your `else` block contains a `return` statement.
that means you do **NOT** need the `else` block, and it is just causing visual clutter.
```
int Type::someFunction()
{
    if( someExpression )
    {
        someStatement();
        return aValue;
    } 
    
    return anotherValue;
}
```
- [ ] revise